Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -633,6 +642,12 @@ google_firestore_v1_Write Serializer::EncodeMutation(
             EncodeFieldTransform(field_transform);
         i++;
       }
+
+      // NOTE: We set a precondition of exists: true as a safety-check, since we
```

Comment:
Note: I think you originally omitted this when implementing mutations in `Serializer`. Integration tests work without it, but perhaps it should be restored for consistency with [Objective-C](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/Source/Remote/FSTSerializerBeta.mm#L536-L539)?

Reply:
Great catch! This omission was a serious oversight.

The integration tests work probably because there isn't a test that verifies this is done, but it's very important because it prevents us from violating the promise that patches won't create partial documents. We should add a test that verifies that patching a document that doesn't exist 

However, one thing this shows is that nanopb makes us very sloppy: `current_document` should only be set if we have a precondition to send, but we're sending it unconditionally. Also, don't we have to set `which_condition_type`? We definitely need a test that this works.

Reply:
Based on our offline discussion, the error case seems impossible to trigger, so we decided against adding the test.